scipy/linalg/_backend.py

```diff
@@ -0,0 +1,60 @@
+import scipy._lib.uarray as ua
+from scipy.linalg import _api
+
+
+__all__ = ['register_backend', 'set_backend',
+           'set_global_backend', 'skip_backend']
+
+
+class _ScipyLinalgBackend:
+    __ua_domain__ = "numpy.scipy.linalg"
+
+    @staticmethod
+    def __ua_function__(method, args, kwargs):
+        fn = getattr(_api, method.__name__, None)
+
+        if fn is None:
+            return NotImplemented
+        return fn(*args, **kwargs)
+
+
+_named_backends = {
+    'scipy': _ScipyLinalgBackend,
```

Comment:
What are the other backends here? Are we preparing for things that don't exist? What would be the protocol architecture for this?

Reply:
The user can pass any object in that implements the uarray protocol. This string version just means we don't clutter up the namespace more than necessary.

Reply:
cupy, dask, jax all have some subset of linalg implemented.

Reply:
As far as I can see they override the functions but it's not like they brought a new blas/lapack alternative as you would for fft. In fact yhis can be done via `scipy.func = cupy.func` with no need for any ua stuff.

Reply:
> In fact yhis can be done via `scipy.func = cupy.func` with no need for any ua stuff.

It may sound trivial but yeah this is also an easy option. Whenever a user registers a backend, just fetch the method from that namespace. For example, `getattr(backend, method)`, where backend can be `backend = scipy`, `backend = cupy`. We may need a conversion layer in between to convert numpy to cupy (for cupy backend) arrays but I think that's it? I am sure that I might be missing something here and would like to know what it is?

Reply:
> As far as I can see they override the functions but it's not like they brought a new blas/lapack alternative as you would for fft. 

I don't think I follow the difference? CuPy's linalg functions are calling cuBLAS/cuSOLVER, so it is an alternative implementation. I don't think monkey-patching is an equivalent solution and it seems like it would lead to code that is harder to debug.

The uarray approach also allows array coercion (if desired) and potentially falling back to a different registered backend in the case where a given function is unavailable in the first backend that was tried. For example, pyFFTW's 0.12 release did not implement the real-to-real FFTs in its backend, so those would end up falling back to SciPy's implementation.

Reply:
Monkeypatching seems undesired but definitely not more complicated than all this at all. You maintain a single list of overloaded functions and you are done. 

What I'm hesitating about is that this is a developer-feature and only a handful of libraries will ever bother with it. And they can monkey-patch 20-30 functions without any problems. Like you said, if there is a more capable FFTW is available and we fall back to pocketfft or whatever the fallback is then it is clear. So a user is deciding to use this or that. 

But here it is not the case. This is something for dask, cupy, etc. users for a drop in replacement. And I don't see why a simple monkeypatching is not enough for those libraries which they implemented almost all already. 

So the concrete question is why are we adding this extra layer of uarray complexity instead of the current monkey patching? It worked until now and will create extra work when we decide to change the signatures or other details internal to scipy. 

You might say no that is not necessary every library can do it themselves then we don't need all this in the first place.

Reply:
>So the concrete question is why are we adding this extra layer of uarray complexity instead of the current monkey patching? It worked until now and will create extra work when we decide to change the signatures or other details internal to scipy.

Are you suggesting something like this?
```python
old_scipy_func = scipy.linalg.func
def patched_func(a):
    if isinstance(a, cupy.ndarray):
        return cupyx.scipy.linalg.func(a)
    else:
        return old_scipy_func(a)
    
scipy.linalg.func = patched_func
```

The biggest problem IMO is that SciPy's signature is being hijacked by the monkey-patcher. That would make it truly difficult to change signatures.

Reply:
That's my worry too indeed. But I don't see how we circumvent it via `uarray` since the replacing backend can still have a different signature and they maintain their own signatures. In both cases the caller has to maintain the correctness of the signature.

Reply:
>That's my worry too indeed. But I don't see how we circumvent it via uarray since the replacing backend can still have a different signature and they maintain their own signatures. 

No, with `uarray` SciPy controls the function arguments 100%. Arguments that SciPy doesn't support will fail to pass through the argument extractor (the function decorated with `ua.create_multimethod`), meaning backend libraries can't change SciPy's argument list. 

SciPy can also add new default arguments without issue. For NumPy arrays, those extra arguments should pass through the dispatch process without touching CuPy's version of that function so it all works. `uarray` even removes the default arguments so CuPy doesn't break for code that worked previously.

Reply:
Then again that is also an answer to your monkey-patching signature concern that we haven't solved anything extra with uarray. Is that a fair assessment?

Reply:
> Then again that is also an answer to your monkey-patching signature concern that we haven't solved anything extra with uarray. Is that a fair assessment?

I don't think it is. What we want is retain the ability to extend the signature in a backwards compatible way (e.g., add a new keyword) without breaking anything downstream. This is possible with what Peter explains. Which is good, otherwise that would be an extra restriction on adding new functionality to SciPy.

> The biggest problem IMO is that SciPy's signature is being hijacked by the monkey-patcher

I don't think that's the biggest issue. Monkeypatching is a big no-no for a _library_. If CuPy monkeypatched SciPy, the first thing we would do is open a bug report. An _end user_ may monkeypatch something if _really_ needed, but it's cumbersome and fragile.

- From a library author perspective the choices are: have a proper dispatch system, or do nothing.
- From an end user perspective the choices are: things work transparently with a single API if there is a dispatch system,  or just use different API's (monkypatching is a lot worse than just modifying all the imports).

I think your questions are valid @ilayn, and it shows that more explanation/rationale is needed of uarray machinery (and uarray itself still needs more gentle intro docs). The monkeypatching discussion isn't relevant though, monkeypatching is a nonstarter no matter how you look at it.

Reply:
What I am trying to emphasize is that uarray brings no convenience if I already use any of the aforementioned libraries. Examples go a long way I think ; 

FFTW vs PocketFFT or MKL vs OpenBLAS is a big deal in terms of backends. That is perfectly fine and I really don't mind if uarray or any other consolidated effort goes into every module. 

Here I am sitting with SciPy and say CuPy installed and imported. Now I'm changing linalg backend to CuPy and want to use eigh. What is the point of using `scipy.linalg.eigh` instead of `cupy.linalg.eigh`? I haven't understood this very point. 

You might argue that well we supply the same code to both cases and one goes to scipy BLAS/LAPACK and the other goes some other whatever. That I sincerely cannot see and to me this seems goldplating relative imports ala

```python
# from scipy.linalg import eigh 
cupy.linalg import eigh
```

That's what I meant by monkey-patching and looks like way easier than all this orchestration of n libraries separately.

Reply:
The end goal is to enable code that work for all array APIs. So for example you could write a library once and have users run it on GPUs with cupy without needing to copy-paste the entire library to change the imports.

Even if you do control the whole stack of code, maybe you'd like to use CUDA when available but still need CPU support. It's not scalable to uncomment all the right lines.

Reply:
For arrays, I totally agree and I think there is lots of discussion about NEPs there but this is public functions we are discussing here. Again, examples are important, 

```
if gpu_condition:
    import cupy as scipy
else:
    import scipy
```

is still the very same thing. I am trying to get to a meaningful example that would demonstrate the benefit of all this. Abstract discussions are nice but at some point they have to be projected onto actual code. Not claiming uarray is not useful but it needs to be fleshed out why we are doing this. So far it looks like a avoidable abstraction layer.

Reply:
To make Peter's example more concrete:

Author 1 writes a package that depends on SciPy (could be a small new package, could be (e.g. scikit-learn) and uses:
```
from scipy import linalg

def somefunc(data, kernel):
    s, v, d = linalg.svd(data)
    z = .... # use svd, kernel
    return z
```
End user 1 uses that package and SciPy. They cannot go in and modify the above package, which comes from PyPI. But they do want to write:
```
from scipy import linalg, fft, ndimage
from package_above import somefunc
import cupy as cp

mydata = cp.load('mydata.npy')
kernel = cp.eye(7)
somefunc(mydata, kernel)
...
```

The end user doesn't want to go talk to the package author first about please using an if-else import. The if-else pattern works only for a single package and end users of just that package, it doesn't scale beyond that.

We're working on preparing a high-level overview (this discussion doesn't belong in a PR comment) and @AnirudhDagar is working on a demo with a nontrivial real-world use case. The `ndimage` and `linalg` backend patterns follow what `fft` already does; @AnirudhDagar and @czgdp1807 are just going quite fast and the rest of the story isn't quite ready yet.

Reply:
So basically we are making changes in the root so that whatever packages are built using that root, do not have to switch to different backends using `if-else` approach. One can call, `scipy.linalg.svd(...)` in their package (say A). If the user of package A passes a cupy array then developer of A won't have to worry because `scipy.linalg.svd` would automatically dispatch the correct implementation, right?

Reply:
Indeed. 

We want to get to a point where pure Python package authors don't have to worry about anything, they can just build on top of NumPy/SciPy - and then for the end user things work well with other compatible libraries like CuPy. At least for a well-defined subset of NumPy and SciPy APIs (clearly making it work for every last function would be too ambitious/unrealistic).

Reply:
Then I think this is one of the most important advantages offered by `uarray`. One can definitely do `if-else` checks to import the right backend and use it but that would make life difficult for developers who use, NumPy/SciPy as dependencies. Moreover, by using only SciPy API calls, one can use different hardwares by just changing the backend. One would not need to use CuPy calls for GPU, SciPy calls for CPU, etc. This introduces uniformity in the user codes.

Reply:
You have to select the backend anyways instead of the import you have to if-else the backends internally. But continuing from that example, one function would be trying dask stuff the other one is going to be trying to import from cupy etc. 

The main issue is that we are trying to become the adapter between End User 1 and Author 1 which is happening outside our jurisdiction but somehow we provide the low-level implementation machinery selection. Author1 doesn't foresee the use of CuPy but we are monkeypatching `package_above`. 

Also this SO says a lot about the adoption of FFT backend possibilities https://stackoverflow.com/questions/62465583/scipy-fft-backends-possibilities hence my comment. 

If we are not really ready can we defer this after everything is in place? Because this all seems a bit premature and overengineered. I am not sure if linalg is the right choice for future experimentation.

Reply:
> You have to select the backend anyways instead of the import you have to if-else the backends internally. 

AFAICT from previous observations, just passing CuPy arrays to SciPy APIs will dispatch CuPy implementation of the SciPy function. So, one may avoid setting the backend and it would be automatically handled by SciPy. I am not sure about this. @AnirudhDagar can confirm.

Reply:
It is still under discussion. See issue #14266. But, at the moment auto-dispatch is not possible.

The easiest way to use is to register the `cupy` backend once at the top, eg: `scipy.fft.register_backend(cu_fft)`. Now the user can use both `cupy` as well `scipy` backends without changing anything and will be able to avoid context manager. Based on the input array, `uarray` will choose the correct backend.

Reply:
> If we are not really ready can we defer this after everything is in place? 

I'd suggest we can/should defer merging it till the bigger-picture docs are ready and also we've had a mailing list discussion. I don't want to see it developed all somewhere away from the SciPy repo though - a WIP PR seems like a good place to work on this.

Reply:
@ilayn I think with the following set of docs/discussions we have sketched the high level picture here:
- https://labs.quansight.org/blog/2021/11/pydata-extensibility-vision/
- https://discuss.scientific-python.org/t/a-proposed-design-for-supporting-multiple-array-types-across-scipy-scikit-learn-scikit-image-and-beyond/131/4
- https://discuss.scientific-python.org/t/default-dispatching-behavior-for-supporting-multiple-array-types-across-scipy-scikit-learn-scikit-image/135
- https://mail.python.org/archives/list/scipy-dev@python.org/thread/RJ7P6ZMW66ZLKSY36RDOCXV3ODHTXLMM/

If you think there is still important things missing or unclear in the content in those links, now would be a good time to respond to one of those threads.